src/FSharpPlus/Data/Seq.fs

```diff
@@ -16,73 +16,885 @@ module Seq =
+                                          x.MoveNext ())
+                              | TryWithState.HaveBodyEnumerator e ->
+                                  let res = ref Unchecked.defaultof<Choice<bool, exn>>
+                                  try
+                                      let! r = e.MoveNext ()
+                                      res := Choice1Of2 r
+                                  with exn ->
+                                      res := Choice2Of2 exn
+                                  match res.Value with
+                                  | Choice1Of2 res ->
+                                      return
+                                          (match res with
+                                           | false -> x.Dispose ()
+                                           | true  -> current := Some e.Current
+                                           res)
+                                  | Choice2Of2 exn ->
+                                      return! (
+                                         x.Dispose ()
+                                         let e = (handler exn :> IEnumerableM<'``Monad<bool>``, 'T>).GetEnumerator ()
+                                         state := TryWithState.HaveHandlerEnumerator e
+                                         x.MoveNext ())
+                              | TryWithState.HaveHandlerEnumerator e ->
+                                  let! res = e.MoveNext()
+                                  return (
+                                      if res then
+                                          current := Some e.Current
+                                          true
+                                      else
+                                          x.Dispose ()
+                                          false)
+                              | _ ->
+                                  return false }
+                                
+                          member _.Dispose () =
+                              match !state with
+                              | TryWithState.HaveBodyEnumerator e | TryWithState.HaveHandlerEnumerator e ->
+                                  state := TryWithState.Finished
+                                  dispose e
+                              | _ -> () } }
 
-    [<EditorBrowsable(EditorBrowsableState.Never)>]
-    static member inline Lift2 (f: 'T->'U->'V, x: SeqT<'``Monad<seq<'T>``>, y: SeqT<'``Monad<seq<'U>``>) = SeqT.lift2 f x y : SeqT<'``Monad<seq<'V>``>
 
-    [<EditorBrowsable(EditorBrowsableState.Never)>]
-    static member inline Lift3 (f: 'T->'U->'V->'W, x: SeqT<'``Monad<seq<'T>``>, y: SeqT<'``Monad<seq<'U>``>, z: SeqT<'``Monad<seq<'V>``>) = SeqT.lift3 f x y z : SeqT<'``Monad<seq<'W>``>
+    [<RequireQualifiedAccess>]
+    type TryFinallyState<'``Monad<bool>``, 'T> =
+       | NotStarted    of SeqT<'``Monad<bool>``, 'T>
+       | HaveBodyEnumerator of IEnumeratorM<'``Monad<bool>``, 'T>
```

Comment:
Naming, consider `HasBodyEnumerator` perhaps (same on others)?

Reply:
> Merely some questions ;). Not yet an in-depth review ;)

Makes sense, you can't review if things aren't clear.